Create 1wf_06_database.md

F-Loyer opened the pull request:
A guide aboute database usage with Caqti and ppx_rapper

F-Loyer committed:
Create 1wf_06_database.md

A guide aboute database usage with Caqti and ppx_rapper

cuihtlauac commented:
Thanks, @F-Loyer, this is awesome. @sabine should this be a guide or cookbook entry?

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+The OCaml ecosystem provides multiple libraries which can be used to access a RDBMS.
+
+One of these libraries is `caqti` which has multiple advantages:
```
```suggestion
One of these libraries is `caqti`. It has the following features:
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+## Introduction
+
+The OCaml ecosystem provides multiple libraries which can be used to access a RDBMS.
```
```suggestion
The OCaml ecosystem provides multiple libraries that can be used to access an RDBMS.
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+- it can be used in a Lwt context which permits its use in some application server. Especially
+the Dream framework which has some specific functions which helps using Caqti,
+- it can be used with the `ppx_rapper` preprocessor. This preprocessor converts an annoted
+SQL query into a query function which leverage the strong typing feature of OCaml.
```
```suggestion
- It can be used with multiple RDBMS (Postgres, MariaDB, SQLite3).
- It can be used in a Lwt context. That permits use in some application servers. Especially the Dream framework, that has some dedicated functions that help using Caqti.
- It can be used with the `ppx_rapper` preprocessor. This preprocessor converts an annotated SQL query into a query function that leverages the strong typing feature of OCaml.
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+Chaining multiple queries may be cumbersome: testing the result value, extracting the
+"expected result" if Ok... The `Lwt_result` monad can make implicit each error test and
+result extraction.
```
```suggestion
Chaining multiple queries may be cumbersome: testing the result value and extracting the expected result if `Ok`. The `Lwt_result` monad can make implicit each error test and result extraction.
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+"expected result" if Ok... The `Lwt_result` monad can make implicit each error test and
+result extraction.
+
+The idea is instead of presenting the program like:
```
```suggestion
The idea is instead of presenting the program like this:
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
````diff
@@ -0,0 +1,328 @@
+  ...
+```
+
+We present the program like:
````
```suggestion
We present the program like that:
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+open Lwt_result.Syntax
+
+let () =
+  match Lwt_main.run(
```
```suggestion
  match Lwt_main.run (
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+(a `query1 ()` is typically of type `('a, 'b) result Lwt.t` which represent a computation
+which need to be scheduled). Once the query has been scheduled and executed, the `Result`
+value is analysed and the value of `Result.Ok value` is extracted. If the result is `Error err`,
+the thread of queries is aborted.
```
```suggestion
The `let*` notation is associated with the `bind` function that deals with Lwt sequencing. A `query1 ()` has often type `('a, 'b) result Lwt.t`, which represents a computation that needs to be scheduled. Once the query has been scheduled and executed, the `Result` value is tested. If the value is available in `Ok`, it is extracted. If the result is `Error`, the thread of queries is aborted.
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+which need to be scheduled). Once the query has been scheduled and executed, the `Result`
+value is analysed and the value of `Result.Ok value` is extracted. If the result is `Error err`,
+the thread of queries is aborted.
+
```
```suggestion
Refer to the tutorials on [Monads](/docs/monads) and [operators](/docs/operators) for more on bind operators and custom lets operators.

```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+## The database connection
+
+`caqti` provides a `connect` function which establish a connection to the database, but it
+is more secure to use `with_connection` which garantee the closure of the database even in the
+presence of an error.
```
```suggestion
`caqti` provides a `connect` function that establishes a connection to the database. Still, it is more secure to use `with_connection`, which ensures the database's closure even in the presence of an error.
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
````diff
@@ -0,0 +1,328 @@
+                                    WHERE category = %string{category} |sql} ]  
+```
+
+`%type{name}` is used to indicate parameters which will be provided when calling the query.
+`@type{name}` is uses to indicate values (column names) which will be returned by the query.
````
To me, this hard to understand.

christinerose replied:
```suggestion
`@type{name}` is used to indicate values (column names), which will be returned by the query.
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
````diff
@@ -0,0 +1,328 @@
+
+```ocaml
+let query = [%rapper get_many {sql| SELECT @int{id}, @string?{value} FROM a
+                                    WHERE category = %string{category} |sql} ]  
````
```suggestion
                                    WHERE category = %string{category} |sql} ]
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+There are different types of queries. A `get_many` query (like the example) returns a list of
+rows. `get_one` just returns the unique row, `get_opt` returns an option value (`None` or `Some`
+depending of the existance of a queried row), `execute` returns nothing `()`.
```
```suggestion
depending on the existence of a queried row), `execute` returns nothing `()`.
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+rows. `get_one` just returns the unique row, `get_opt` returns an option value (`None` or `Some`
+depending of the existance of a queried row), `execute` returns nothing `()`.
+
+By default rows are presented as tuples.
```
```suggestion
By default, rows are presented as tuples.
```

Is “presented” the right verb?

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+Where supported options are:
+
+- `record_in` : instead of calling the query with one argument per parameter, a single
+argument is provided. It must be a record whose the fields correspond to the parameters.
```
```suggestion
argument is provided. It must be a record whose fields correspond to the parameters.
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+## Itering a list of queries, transactions
+
+Using `List.map` can easily create queries... but they won't be chained from
```
```suggestion
Using `List.map` creates queries that won't be chained from
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+## Itering a list of queries, transactions
+
+Using `List.map` can easily create queries... but they won't be chained from
+a `lwt` point of vue. The `Lwt_list.iter_s` and `Lwt_list.map_s` can iter
```
```suggestion
a `lwt` point of view. The `Lwt_list.iter_s` and `Lwt_list.map_s` can iter
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+them... but won't stop at the first error.
+
+If we wan't to iter all the queries even in the presence of error, we can
+simply type
```
```suggestion
If we want to iter all the queries even in the presence of an error, we can
type
```

christinerose replied:
I'd also put a comma before "even"

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+This function iter a list of functions with a connexion as an argument and execute each query.
+It can be used in the following context (where `insert_a_record` is defined in the example
+section):
```
```suggestion
This function iters a list of functions with a connection as an argument and executes each query.
It can be used in the following context (where `insert_a_record` is defined in the example
section):
```

cuihtlauac commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+## Example
+
+The following example is a working program which illustrates the use of the different
```
```suggestion
The following example is a working program that illustrates the use of the different
```

cuihtlauac reviewed (changes requested):
- A couple of complicated sentences
- Forbidden words: simple, easy

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+This tutorial is made to facilitate the integration of the multiple involved libraries.
+
+## The Lwt and Lwt_result context
```
```suggestion
## The Lwt and `Lwt_result` context
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+## The Lwt and Lwt_result context
+
+`caqti` provides blocking functions which avoid the need of `lwt`, however one can use the `ppx_rapper`
```
```suggestion
`caqti` provides blocking functions, which avoid the need of `lwt`. However, one can use the `ppx_rapper`
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
````diff
@@ -0,0 +1,328 @@
+  | Result.Error r -> ...
+```
+
+The `let*` notation is associated to the `bind` function which deals with Lwt sequencing
````
```suggestion
The `let*` notation is associated to the `bind` function, which deals with Lwt sequencing
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+value is analysed and the value of `Result.Ok value` is extracted. If the result is `Error err`,
+the thread of queries is aborted.
+
+## The database connection
```
```suggestion
## The Database Connection
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+- `"mariadb://user:passwd@host/database"` (using mariadb library),
+
+
+## Declaring and using queries
```
```suggestion
## Declaring and Using Queries
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+## Declaring and using queries
+
+### Default queries
```
```suggestion
### Default Queries
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
````diff
@@ -0,0 +1,328 @@
+                                    WHERE category = %string{category} |sql} ]  
+```
+
+`%type{name}` is used to indicate parameters which will be provided when calling the query.
````
```suggestion
`%type{name}` is used to indicate parameters that will be provided when calling the query.
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+By default rows are presented as tuples.
+
+### ppx_rapper options
```
```suggestion
### `ppx_rapper` Options
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+### ppx_rapper options
+
+Depending of the needs, a `ppx_rapper` query can be written as
```
```suggestion
Depending on the needs, a `ppx_rapper` query can be written as
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+argument is provided. It must be a record whose the fields correspond to the parameters.
+- `record_out` : instead of outputing row(s) as tuple(s), they are presented as record(s).
+- `function_out` : the query is called with a function as its first argument. This function
+is called for each row with every values (column names) as a labelled arguments. The return value of the query
```
```suggestion
is called for each row with every value (column names) as a labelled argument. The return value of the query
```
This either needs to be
"every value" or just "values" 
as well as:
"a labelled argument" or "labelled arguments" (without the "a")

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+- `record_out` : instead of outputing row(s) as tuple(s), they are presented as record(s).
+- `function_out` : the query is called with a function as its first argument. This function
+is called for each row with every values (column names) as a labelled arguments. The return value of the query
+is composed of the returned values of this function. This works like [List/Option.map].
```
```suggestion
is composed of the function's returned values. This works like [List/Option.map].
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
```diff
@@ -0,0 +1,328 @@
+
+Options are separated by whitespaces.
+
+## Custom error handling
```
```suggestion
## Custom Error Handling
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
````diff
@@ -0,0 +1,328 @@
+                  Lwt.return (Result.Error err)
+```
+
+Note: we can also change the `Lwt.return` values... An ok query could be
````
```suggestion
Note: we can also change the `Lwt.return` values... An acceptable query could be
```

christinerose replied:
```suggestion
Note: we can also change the `Lwt.return` values... An `Ok` query could be
```

christinerose commented on data/tutorials/guides/1wf_06_database.md:
````diff
@@ -0,0 +1,328 @@
+```
+
+Note: we can also change the `Lwt.return` values... An ok query could be
+wrong if the returned values do not respect some assertions. An error query could also be mitigated.
````
```suggestion
wrong if the returned values do not respect some assertions. An `Error` query could also be mitigated.
```
Should these match the code in capitalization?

christinerose commented on data/tutorials/guides/1wf_06_database.md:
````diff
@@ -0,0 +1,328 @@
+ let* () = match%lwt ... in
+```
+
+## Itering a list of queries, transactions
````
```suggestion
## Itering a List of Queries, Transactions
```

christinerose reviewed (commented):
Some grammar/formatting suggestions. I agree with everything @cuihtlauac suggested as well.

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

F-Loyer committed:
Update data/tutorials/guides/1wf_06_database.md

Co-authored-by: Christine Rose <christinerose@users.noreply.github.com>

F-Loyer committed:
Update 1wf_06_database.md

Change acceptable to `Ok`. Since the subject is the `Result` match of the query

F-Loyer committed:
Update 1wf_06_database.md

Change easy to easier

F-Loyer committed:
Update 1wf_02_error_handling.md

Typo about map_error

F-Loyer commented:
> * A couple of complicated sentences
> 
>     * Forbidden words: simple, easy

easy -> easier (ppx_rapper does make things easier ! Easy, it may depends of the programmer)
Which complicated sentences should be reworded ?

F-Loyer reviewed (commented):
All changes resolved
